### PR TITLE
Clean up and improve registries

### PR DESCRIPTION
**Pull request by MrIbby**
Changes:
* Documentation
* Rename obfuscated fields
* Add AutomaticNamespacedRegistry and AutomaticDefaultedNamespacedRegistry - the `Item` registry should not have a default key
* Rearrange order of parameters for item and block registration - there is no reason to put the ID last

### CODE REVIEW

File: src/main/java/xyz/openmodloader/registry/OMLDefaultedNamespacedRegistry.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
+package xyz.openmodloader.registry;
+
+import org.apache.commons.lang3.Validate;
+
+import java.util.Random;
+import java.util.function.Function;
+
+public class OMLDefaultedNamespacedRegistry<K, V> extends OMLNamespacedRegistry<K, V> {
```
Comment (phase) [inline src/main/java/xyz/openmodloader/registry/OMLDefaultedNamespacedRegistry.java]: `DefaultNamespacedRegistry` and `NamespacedRegistry` are better. We don't need prefixes.

Comment (MrIbby) [inline src/main/java/xyz/openmodloader/registry/OMLDefaultedNamespacedRegistry.java]: Well, `OMLNamespacedRegistry` extends `RegistryNamespaced`, and it probably would be best not to have `NamespacedRegistry` and `RegistryNamespaced` coexist.

File: src/main/java/xyz/openmodloader/registry/GameRegistry.java [hunk 2]
```diff
@@ -9,70 +8,91 @@
...
-    public static void registerHorseArmor(ResourceLocation identifier, HorseArmorType horseArmor) {
-        OMLRegistry.getRegistry(HorseArmorType.class).register(identifier, horseArmor);
+    public static void registerHorseArmor(ResourceLocation id, HorseArmorType horseArmor) {
+        if (horseArmor == null) {
+            throw new IllegalArgumentException("Attempted to register null horse armor");
+        }
+        if (id == null) {
+            throw new IllegalArgumentException("Attempted to register horse armor with null ID");
+        }
+        OMLNamespacedRegistry<ResourceLocation, HorseArmorType> horseArmorRegistry = OMLRegistry.getRegistry(HorseArmorType.class);
```
Comment (lclc98) [inline src/main/java/xyz/openmodloader/registry/GameRegistry.java]: I got a NPE when running this

Comment (MrIbby) [inline src/main/java/xyz/openmodloader/registry/GameRegistry.java]: Fixed

File: src/main/java/xyz/openmodloader/registry/OMLDefaultedNamespacedRegistry.java (hunk 1 above)
Comment (phase) [inline src/main/java/xyz/openmodloader/registry/OMLDefaultedNamespacedRegistry.java]: Then call it something else. What makes it different than Minecraft's `RegistryNamespaced`?

Comment (MrIbby) [inline src/main/java/xyz/openmodloader/registry/OMLDefaultedNamespacedRegistry.java]: It automatically finds and maps available integer IDs. How about `AutomaticNamespacedRegistry`?

Comment (phase) [inline src/main/java/xyz/openmodloader/registry/OMLDefaultedNamespacedRegistry.java]: Perfect!

**Comment by MrIbby** [comment]
Am I done? I think I am.

**Comment by darkhax** [comment]
You're going to hate @coolsquid lol. His world gen PR is probably going to break this.

**Comment by MrIbby** [comment]
Am I done?

File: patches/net/minecraft/entity/passive/HorseArmorType.java.patch [hunk 3]
```diff
@@ -9,16 +9,14 @@
...
-+    DIAMOND = new HorseArmorType(11, "diamond", "dio");
-+    
++    public static final HorseArmorType NONE = new HorseArmorType(0);
++    public static final HorseArmorType IRON = new HorseArmorType(5, "iron", "meo");
++    public static final HorseArmorType GOLD = new HorseArmorType(7, "gold", "goo");
++    public static final HorseArmorType DIAMOND = new HorseArmorType(11, "diamond", "dio");
+ 
      private final String textureName;
      private final String hash;
-     private final int protection;
```
Comment (lclc98) [inline patches/net/minecraft/entity/passive/HorseArmorType.java.patch]: Why did this get removed?

**Pull request merged**